### PR TITLE
[SPARK-31170][SQL] Spark SQL Cli should respect hive-site.xml and spark.sql.warehouse.dir

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

In Spark CLI, we create a hive `CliSessionState` and it does not load the `hive-site.xml`. So the configurations in `hive-site.xml` will not take effects like other spark-hive integration apps.

Also, the warehouse directory is not correctly picked. If the `default` database does not exist, the `CliSessionState` will create one during the first time it talks to the metastore. The `Location` of the default DB will be neither the value of `spark.sql.warehousr.dir` nor the user-specified value of `hive.metastore.warehourse.dir`, but the default value of `hive.metastore.warehourse.dir `which will always be `/user/hive/warehouse`.

### Why are the changes needed?

fix bug for Spark SQL cli to pick right confs

### Does this PR introduce any user-facing change?

yes, the non-exists default database will be created in the location specified by the users via `spark.sql.warehouse.dir` or `hive.metastore.warehouse.dir`, or the default value of `spark.sql.warehouse.dir` if none of them specified.

### How was this patch tested?

add cli ut

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @dongjoon-hyun @maropu, thanks

**Comment by yaooqinn** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala [hunk 1]
```diff
@@ -55,44 +54,10 @@ private[sql] class SharedState(
...
-      val sparkWarehouseDir = sparkContext.conf.get(WAREHOUSE_PATH)
-      logInfo(s"Setting hive.metastore.warehouse.dir ('$hiveWarehouseDir') to the value of " +
-        s"${WAREHOUSE_PATH.key} ('$sparkWarehouseDir').")
-      sparkContext.hadoopConfiguration.set("hive.metastore.warehouse.dir", sparkWarehouseDir)
-      sparkWarehouseDir
-    }
-  }
-  logInfo(s"Warehouse path is '$warehousePath'.")
-
-  // These 2 variables should be initiated after `warehousePath`, because in the first place we need
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala]: shall we keep this comment? Or we can put `SharedState.loadHiveConfFile` inside `private val (conf, hadoopConf) = {`

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 2]
```diff
@@ -159,6 +160,17 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
     }
   }
 
+  test("Pick spark.sql.warehouse.dir first for Spark Cli if set") {
+    val sparkWareHouseDir = Utils.createTempDir()
+    new File(warehousePath, "metastore_db").delete()
+    runCliWithin(
+      1.minute,
+      Seq("--conf", s"""${StaticSQLConf.WAREHOUSE_PATH.key}=$sparkWareHouseDir"""))(
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: why use three `"""` here?

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 3]
```diff
@@ -159,6 +160,17 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
...
+      """desc database default;""" -> sparkWareHouseDir.getAbsolutePath
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: ditto

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 4]
```diff
@@ -159,6 +160,17 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
...
+    )
+    sparkWareHouseDir.delete()
```
Comment (cloud-fan) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: how do we verify if the warehouse path is picked?

Comment (yaooqinn) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: the hive.xxx.xxx.dir is from the global variable `warehousePath`, if the `sparkWareHouseDir` not picked, the db location of default will be the hive one

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala]: ok, I will move it inside.

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala (hunk 4 above)
Comment (yaooqinn) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: I also checked manually
#### now
```sql
kentyao@hulk  ~/Downloads/spark/spark-3.1.0-SNAPSHOT-bin-v20200317  bin/spark-sql -S --conf spark.sql.warehouse.dir=/Users/kentyao/Downloads/spark/spark-3.1.0-SNAPSHOT-bin-v20200317/data/spark-warehouse
spark-sql> desc database default;
Database Name	default
Comment	Default Hive database
Location	file:/Users/kentyao/Downloads/spark/spark-3.1.0-SNAPSHOT-bin-v20200317/data/spark-warehouse
Owner	public
```
#### 3.x and 2.4
```sql

kentyao@hulk  ~/Downloads/spark/spark-2.4.5-bin-hadoop2.7  bin/spark-sql -S -c spark.sql.warehouse.dir=/Users/kentyao/Downloads/spark/spark-3.1.0-SNAPSHOT-bin-v20200317/data/spark-warehouse
spark-sql> desc formatted default;
Error in query: Table or view 'default' not found in database 'default';
spark-sql> desc database default;
Database Name	default
Description	Default Hive database
Location	file:/user/hive/warehouse
```

Comment (cloud-fan) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: How is it verified?

Comment (yaooqinn) [reply sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: In this test, we will run a spark-sql job to check the output
```
[info] - Pick spark.sql.warehouse.dir first for Spark Cli if set *** FAILED *** (1 minute)
[info]   =======================
[info]   CliSuite failure output
[info]   =======================
[info]   Spark SQL CLI command line: ../../bin/spark-sql --master local --driver-java-options -Dderby.system.durability=test --conf spark.ui.enabled=false --hiveconf javax.jdo.option.ConnectionURL=jdbc:derby:;databaseName=/Users/kentyao/spark/target/tmp/spark-3570629a-23a7-4e2a-b21a-9bce68025228;create=true --hiveconf hive.metastore.warehouse.dir=/Users/kentyao/spark/target/tmp/spark-dbe4dd6a-1129-43f5-83bc-62dd2072b410 --hiveconf hive.exec.scratchdir=/Users/kentyao/spark/target/tmp/spark-9422968c-94d0-4251-b591-30cbcc061845 --hiveconf conf1=conftest --hiveconf conf2=1 --conf spark.sql.warehouse.dir=/Users/kentyao/spark/target/tmp/spark-1219faf2-918f-4ac4-8d28-fa4dd2d24c7e
[info]   Exception: java.util.concurrent.TimeoutException: Futures timed out after [1 minute]
[info]   Executed query 0 "desc database default;",
[info]   But failed to capture expected output "/Users/kentyao/spark/target/tmp/spark-1219faf2-918f-4ac4-8d28-fa4dd2d24c7e" within 1 minute.
```

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by gatorsmile** [comment]
cc @juliuszsompolski

File: sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala [hunk 5]
```diff
@@ -159,6 +160,17 @@ class CliSuite extends SparkFunSuite with BeforeAndAfterAll with Logging {
     }
   }
 
+  test("Pick spark.sql.warehouse.dir first for Spark Cli if set") {
```
Comment (dongjoon-hyun) [inline sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/CliSuite.scala]: Hi, All.
This test case broke all `Hive 1.2` profile Jenkins jobs.

**Comment by dongjoon-hyun** [comment]
The situation looks tricky.
```
.../spark-7c73c224-7624-48c5-95dc-a7a8c3bf41bb (Test case is looking for this)
.../spark-e8eb1838-fde8-4465-ada2-86a2b8f07297 (Spark answer in the cluster.)
```

**Comment by dongjoon-hyun** [comment]
When I run `CliSuite` locally with `hive-1.2`, it seems to okay.  I'll revert this first and recover 4 Jenkins jobs. Please make a new PR and test with `hive-1.2` tag, @yaooqinn .

**Comment by cloud-fan** [comment]
Yea let's revert to unblock the builds. @yaooqinn when you open a new PR, please add [hive-1.2] in the title, so jenkins will test with hive 1.2

**Comment by yaooqinn** [comment]
Thanks, @cloud-fan @dongjoon-hyun for the reverting, I will raise a pr to fix this ASAP. Also, sorry for my mistake

**Comment by dongjoon-hyun** [comment]
Thank you so much, @cloud-fan and @yaooqinn . The jobs are recovered.

**Pull request closed without merging**